system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "subscriber_topic_statistics.hpp"
+#include "system_metrics_collector/constants.hpp"
+#include "system_metrics_collector/metrics_message_publisher.hpp"
+
+using metrics_statistics_msgs::msg::MetricsMessage;
+using std::placeholders::_1;
+
+namespace topic_statistics_collector
+{
+template<typename T>
+SubscriberTopicStatistics<T>::SubscriberTopicStatistics(
+  const std::string & node_name,
+  const rclcpp::NodeOptions & node_options,
+  const std::string & topic_name)
+: rclcpp_lifecycle::LifecycleNode{node_name, node_options},
+  topic_name_(topic_name)
+{
+  rcl_interfaces::msg::IntegerRange positive_range;
+  positive_range.from_value = 1;
+  positive_range.to_value =
+    std::numeric_limits<decltype(rcl_interfaces::msg::IntegerRange::to_value)>::max();
+  positive_range.step = 1;
+
+  rcl_interfaces::msg::ParameterDescriptor descriptor;
+  descriptor.read_only = true;
+  descriptor.integer_range.push_back(positive_range);
+
+  descriptor.description =
+    "The period in milliseconds between each published MetricsMessage.";
+  auto publish_period = declare_parameter(
+    system_metrics_collector::collector_node_constants::kPublishPeriodParam,
+    system_metrics_collector::collector_node_constants::kDefaultPublishPeriod.count(),
+    descriptor);
+  publish_period_ = std::chrono::milliseconds{publish_period};
+
+  statistics_collectors_.push_back(std::make_unique<ReceivedMessageAgeCollector<T>>());
+  statistics_collectors_.push_back(std::make_unique<ReceivedMessagePeriodCollector<T>>());
+
+  auto qos_options_ = rclcpp::QoS(rclcpp::KeepAll());
+  subscription_ = create_subscription<T>(
+    topic_name_,
+    qos_options_,
+    std::bind(&SubscriberTopicStatistics::CollectorCallback, this, _1),
+    subscription_options_);
+}
+
+template<typename T>
+bool SubscriberTopicStatistics<T>::SetupStart()
+{
+  assert(publish_timer_ == nullptr);
```

Comment:
Consider using [rcpputils::check_true](https://github.com/ros2/rcpputils/blob/master/include/rcpputils/asserts.hpp) instead of `assert` since it will throw if the statement is false. This can then be checked in the unit tests.

Reply:
Thanks! Wasn't aware of this header.